protocols/kad/src/handler.rs

```diff
@@ -766,15 +771,24 @@ where
             ));
         }
 
-        if let Poll::Ready(Some(event)) = self.outbound_substreams.poll_next_unpin(cx) {
+        let num_in_progress_outbound_substreams =
+            self.active_outbound_substreams.len() + self.num_upgrading_outbound_streams;
+        if MAX_NUM_SUBSTREAMS.saturating_sub(num_in_progress_outbound_substreams) > 0 {
+            if let Some(protocol) = self.next_outbound_streams.pop_front() {
+                self.num_upgrading_outbound_streams += 1;
+                return Poll::Ready(ConnectionHandlerEvent::OutboundSubstreamRequest { protocol });
+            }
+        }
```

Comment:
I think we should move this further down, i.e. just before returning `Poll::Pending`. That way, we prioritize existing work (polling active substreams) over requesting new ones.

Reply:
No strong opinion on this one, hence moved as suggested